### PR TITLE
Feature/hg 77 creating rooms hpi campus i

### PR DESCRIPTION
**Pull request by hpistudent72**
Added Indoor Maps for ABC-Gebäude and Hauptgebäude. Additionally updated the map for Hörsaalgebäude.

### CODE REVIEW

**Comment by oleschl** [comment]
I don't know how you exported the geojson however your rooms use a LineString instead of a Polygon for the room geometry.

**Comment by oleschl** [comment]
The lines of the rooms of the main building are crooked and look a little messy sometimes. That could be improved but would take some time. @ClFeSc  and @anonym-HPI  what is your opinion on this?
Current:
Screenshot from 2022-01-20 16-08-05

Quick improve attempt:
Screenshot from 2022-01-20 16-18-51

**Comment by hpistudent72** [comment]
> I don't know how you exported the geojson however your rooms use a LineString instead of a Polygon for the room geometry.

I created them just using save as. Oddly enough I can't open the geojsons using polygons in JOSM. Did you use another extension that wasn't mentioned? Either way editing the geojsons to swap LineString for Polygon shouldn't take long, but I can't  use them afterwards if there are points we have to move.

**Review by ClFeSc — CHANGES_REQUESTED** [review]
I tried importing the geojson files in order to see them, but it fails for me for all buildings but the ABC building with some logs like `feature has no valid level name and will be skipped:  gorund` and `Uncaught (in promise) Error: Invalid LatLng object: (NaN, NaN)`. Please add a working method of import to the branch so that the content can actually be used.

**Comment by ClFeSc** [comment]
> The lines of the rooms of the main building are crooked and look a little messy sometimes. That could be improved but would take some time. @ClFeSc and @anonym-HPI what is your opinion on this? Current: Screenshot from 2022-01-20 16-08-05
> 
> Quick improve attempt: Screenshot from 2022-01-20 16-18-51

I don't really like the offset, but I don't think it's worth the hassle and the time improving it. Unless @anonym-HPI says it should be done.

File: app/assets/geojsons/HG_-1.geojson [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                    ],
+                    [
+                        13.13219039206,
+                        52.39401088823
+                    ],
+                    [
+                        13.13219416721,
+                        52.39397398336
+                    ]
+                ]
```
Comment (Dassderdie) [inline app/assets/geojsons/HG_-1.geojson:42]: 1. to see see the rooms one has to add 
```js
  await loadGeoJsonFile("/assets/H_0.geojson");
  await loadGeoJsonFile("/assets/H_1.geojson");
  await loadGeoJsonFile("/assets/H_2.geojson");
  await loadGeoJsonFile("/assets/H_3.geojson");
  await loadGeoJsonFile("/assets/HG_-1.geojson");
  await loadGeoJsonFile("/assets/HS_0.geojson");
```
into `leafletMap.js`.
(or the more performant way)

```js
    await Promise.all(
        loadGeoJsonFile('/assets/ABC-Building-0.geojson'),
        loadGeoJsonFile('/assets/ABC-Building-1.geojson'),
        loadGeoJsonFile('/assets/H_0.geojson'),
        loadGeoJsonFile('/assets/H_1.geojson'),
        loadGeoJsonFile('/assets/H_2.geojson'),
        loadGeoJsonFile('/assets/H_3.geojson'),
        loadGeoJsonFile('/assets/HG_-1.geojson'),
        loadGeoJsonFile('/assets/HS_0.geojson')
    );
```.

Then the js errors with:
```log
leaflet-src.js:1280 Uncaught (in promise) Error: Invalid LatLng object: (undefined, undefined)
    at new LatLng (leaflet-src.js:1280:1)
    at coordsToLatLng (leaflet-src.js:8320:1)
    at coordsToLatLngs (leaflet-src.js:8331:1)
    at coordsToLatLngs (leaflet-src.js:8331:1)
    at geometryToLayer (leaflet-src.js:8289:1)
    at NewClass.addData (leaflet-src.js:8200:1)
    at NewClass.addData (leaflet-src.js:8187:1)
    at NewClass.initialize (leaflet-src.js:8170:1)
    at new NewClass (leaflet-src.js:296:1)
    at Object.geoJSON (leaflet-src.js:8502:1)
```

It seems to me like this is caused by incorrect Polygon coordiantes.
In the (working ABC geojsons there is another array around the coordinates in the polygon):
```suggestion
                "coordinates": [[
                    [
                        13.13219416721,
                        52.39397398336
                    ],
                    [
                        13.13208460674,
                        52.39396990756
                    ],
                    [
                        13.13207924232,
                        52.39402146549
                    ],
                    [
                        13.13218842279,
                        52.39402490812
                    ],
                    [
                        13.13219039206,
                        52.39401088823
                    ],
                    [
                        13.13219416721,
                        52.39397398336
                    ]
                ]]
```

Comment (Dassderdie) [reply app/assets/geojsons/HG_-1.geojson:42]: In TS:

```ts
// currently:
type Coordinates = [number, number][]
// correct:
type Coordinates = [[number, number][]]

```

File: app/javascript/packs/leafletMap.js [hunk 2]
```diff
@@ -27,263 +27,269 @@ let currentFloor = 0;
...
+
+    // find a position, create an invisible marker and bind a tooltip to it
+    // we do not bind the tooltip to the room because you can only change the position of a tooltip by defining an offset
+    // to the referenced object (you cannot give it a coordinate), we would need to determine an offset for each room
+
+    // used for polylabel to estimate the visual center of a polygon that is inside of the polygon
+    // lower value -> higher precision
+    const markerPositionPrecision = 0.000001;
+    const markerPosition = polylabel(
+        feature.geometry.type == "Polygon"? feature.geometry.coordinates : feature.geometry.coordinates[0],
```
Comment (ClFeSc) [inline app/javascript/packs/leafletMap.js:255]:
```suggestion
        feature.geometry.type === "Polygon" ? feature.geometry.coordinates : feature.geometry.coordinates[0],
```

File: app/javascript/packs/leafletMap.js [hunk 3]
```diff
@@ -27,263 +27,269 @@ let currentFloor = 0;
...
+    addPolygons(buildingPolygons);
+    addMarkers(buildingMarkers);
+
+    // display indoor information eg. rooms, labels
+    await loadGeoJsonFile("/assets/ABC-Building-0.geojson");
+    await loadGeoJsonFile("/assets/ABC-Building-1.geojson");
+    await loadGeoJsonFile("/assets/ABC-Building-2.geojson");
+    await loadGeoJsonFile("/assets/H_0.geojson");
+    await loadGeoJsonFile("/assets/H_1.geojson");
+    await loadGeoJsonFile("/assets/H_2.geojson");
```
Comment (ClFeSc) [inline app/javascript/packs/leafletMap.js:58]:
```suggestion
    await loadGeoJsonFile("/assets/H_2.geojson");
    await loadGeoJsonFile("/assets/H_3.geojson");
```

File: app/javascript/packs/leafletMap.js [hunk 4]
```diff
@@ -27,263 +27,269 @@ let currentFloor = 0;
...
-    const newLayer = L.layerGroup([floors[level].rooms, floors[level].labels]);
-    layerControl.addBaseLayer(newLayer, feature.properties.level_name);
-    // We add the current floor to the map here so that the map and layer control reference the same object
-    // now the layer control will select the correct check box automatically
-    if (level === currentFloor) {
-      newLayer.addTo(map);
+    if (!floors[level]) {
+        floors[level] = { rooms: L.layerGroup(), labels: L.layerGroup() };
+        const newLayer = L.layerGroup([floors[level].rooms, floors[level].labels]);
+        layerControl.addBaseLayer(newLayer, feature.properties.level_name);
```
Comment (ClFeSc) [inline app/javascript/packs/leafletMap.js:237]: Could we somehow manage to order these layers? The order currently is `0` -> `1` -> `2` -> `-1`.

**Review by ClFeSc — CHANGES_REQUESTED** [review]
I don't know whether the calculation of the label/marker positions is correct here?
Bildschirmfoto vom 2022-01-25 20-17-59

Other than that, just some minor points.

File: app/javascript/packs/leafletMap.js [hunk 5]
```diff
@@ -27,263 +27,269 @@ let currentFloor = 0;
...
+
+    // display indoor information eg. rooms, labels
+    await loadGeoJsonFile("/assets/ABC-Building-0.geojson");
+    await loadGeoJsonFile("/assets/ABC-Building-1.geojson");
+    await loadGeoJsonFile("/assets/ABC-Building-2.geojson");
+    await loadGeoJsonFile("/assets/H_0.geojson");
+    await loadGeoJsonFile("/assets/H_1.geojson");
+    await loadGeoJsonFile("/assets/H_2.geojson");
+    await loadGeoJsonFile("/assets/HS_0.geojson");
+    await loadGeoJsonFile("/assets/HG_-1.geojson");
```
Comment (Dassderdie) [inline app/javascript/packs/leafletMap.js:60]: Use `Promise.all()` to parallelize the loading of the geojsons.

**Review by ClFeSc — COMMENTED** [review]
I'm now happy with it, but as I changed some things, I like another one (@Dassderdie maybe) to take another look to approve.

**Comment by ClFeSc** [comment]
I don't know why the test fails, the tested behaviour works for me as far as I can tell.

**Comment by anonym-HPI** [comment]
> The lines of the rooms of the main building are crooked and look a little messy sometimes. That could be improved but would take some time. @ClFeSc and @anonym-HPI what is your opinion on this? Current: Screenshot from 2022-01-20 16-08-05
> 
> Quick improve attempt: Screenshot from 2022-01-20 16-18-51

Sorry, just saw it now. If you mean the entrance (at the door), that is not important, but the pink lines behind of the building make it look not good. A way, would be to hide them in indoor mode, if putting them exactly on top, is a problem. But they should not be too far off, to be noticed when zooming in and out (and then the building "jumps" a bit in a direction or deforms).

**Review by Dassderdie — APPROVED** [review]
LGTM

**Comment by ClFeSc** [comment]
What should we do about the CodeFactor? It only mentions files not modified in this PR.

**Comment by ClFeSc** [comment]
> What should we do about the CodeFactor? It only mentions files not modified in this PR.

I have just clicked on 'Fix that' in the CodeFactor menu.

**Review by ClFeSc — APPROVED** [review]
Should be fine (I hope so).

**Comment by JulianEgbert** [comment]
Please notify our team (T.O.B.I.A.S) via Discord when this is merged.
We can then extend our indoor routing to these buildings.

**Review by anonym-HPI — APPROVED** [review]
visual review approved (descriptions will be hidden in name tag in a seperate issue)

**Pull request merged**